Aluno: Bruno Ricardo Pereira dos Santos Santos

Nº USP: 10288640

In [ ]:
# Importando bibliotecas
import pandas as pd
import string
import nltk
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('rslp')
from nltk.tokenize import word_tokenize
from nltk.stem.porter import *
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from scipy.spatial.distance import cosine
import numpy as np
import networkx as nx
!pip install plotly.express
from plotly import graph_objs as go

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.model_selection import train_test_split

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package rslp to /root/nltk_data...
[nltk_data]   Package rslp is already up-to-date!


 # Identificação do problema

A base usada contém o texto de um SMS e a sua descrição, se é um spam ou não. O objetivo deste projeto é analisar e criar um modelo para classificação de textos de SMS.


In [ ]:
dataset = pd.read_excel('/content/drive/MyDrive/graduação/mineracao_dados_nao_estruturados/SMSSpamCollection.xlsx')
dataset

dataset['text'] = dataset['text'].astype(str)

# Pré-processamento

## Modelo bag-of-words

In [ ]:
# remoção de pontuacao e stopwords

def remove_stopwords(text,lang,domain_stopwords=[]):
  
  stop_words = nltk.corpus.stopwords.words(lang) # lang='portuguese' or lang='english'
  
  s = str(text).lower() # tudo para caixa baixa
  table = str.maketrans({key: None for key in string.punctuation})
  s = s.translate(table) # remove pontuacao
  tokens = word_tokenize(s) #obtem tokens
  v = [i for i in tokens if not i in stop_words and not i in domain_stopwords and not i.isdigit()] # remove stopwords
  s = ""
  for token in v:
    s += token+" "
  return s.strip()

# stemming
def stemming(text,lang):
  
  stemmer = PorterStemmer() # stemming para ingles
  
  if lang=='portuguese':
    stemmer = nltk.stem.RSLPStemmer() # stemming para portuguese
    
  tokens = word_tokenize(text) #obtem tokens
  
  sentence_stem = ''
  doc_text_stems = [stemmer.stem(i) for i in tokens]
  for stem in doc_text_stems:
    sentence_stem += stem+" "
    
  return sentence_stem.strip()

In [ ]:
# Bag of Words com refinamento: ponderação por TF-IDF
def compute_vsm_tfidf(dataset,lang,domain_stopwords=[]):
  
  d = []
  for index,row in dataset.iterrows():
    text = row['text'] #texto do evento
    text2 = remove_stopwords(text, lang,domain_stopwords)
    text3 = stemming(text2, lang)
    d.append(text3)
  matrix = TfidfVectorizer()
  X = matrix.fit_transform(d)
  
  tfidf_vect_df = pd.DataFrame(X.todense(), columns=matrix.get_feature_names())

  return tfidf_vect_df


vsm = compute_vsm_tfidf(dataset,'english')
vsm

0089mi  0776xxxxxxx  077xxx  0789xxxxxxx  ...  ã¼ll   㜠 ãœll  œharri
0        0.0          0.0     0.0          0.0  ...   0.0  0.0   0.0     0.0
1        0.0          0.0     0.0          0.0  ...   0.0  0.0   0.0     0.0
2        0.0          0.0     0.0          0.0  ...   0.0  0.0   0.0     0.0
3        0.0          0.0     0.0          0.0  ...   0.0  0.0   0.0     0.0
4        0.0          0.0     0.0          0.0  ...   0.0  0.0   0.0     0.0
...      ...          ...     ...          ...  ...   ...  ...   ...     ...
5569     0.0          0.0     0.0          0.0  ...   0.0  0.0   0.0     0.0
5570     0.0          0.0     0.0          0.0  ...   0.0  0.0   0.0     0.0
5571     0.0          0.0     0.0          0.0  ...   0.0  0.0   0.0     0.0
5572     0.0          0.0     0.0          0.0  ...   0.0  0.0   0.0     0.0
5573     0.0          0.0     0.0          0.0  ...   0.0  0.0   0.0     0.0

[5574 rows x 7593 columns]

## Word2Vec

Tendo em vista que nossa base de dados não apresenta em sua maior parte: 

- Textos muito complexos
- Palavras com mais de um sentido
- Palavras que invertem o sentido da sentença

Não é necessário usar uma Word Embedding mais robusta como o ELMO ou o BERT para obter um resultado satisfatório, podemos usar uma word Embedding estática. 

In [ ]:
!wget -P /root/input/ -c "https://s3.amazonaws.com/dl4j-distribution/GoogleNews-vectors-negative300.bin.gz"

--2021-11-02 23:04:34--  https://s3.amazonaws.com/dl4j-distribution/GoogleNews-vectors-negative300.bin.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.89.37
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.89.37|:443... connected.
HTTP request sent, awaiting response... 416 Requested Range Not Satisfiable

    The file is already fully retrieved; nothing to do.



In [ ]:
!pip install gensim
from gensim.models import KeyedVectors

In [ ]:
EMBEDDING_FILE = '/root/input/GoogleNews-vectors-negative300.bin.gz' # from above
word2vec = KeyedVectors.load_word2vec_format(EMBEDDING_FILE, binary=True)

In [ ]:
df_w2v_train, df_w2v_test = train_test_split(dataset, test_size=0.30, random_state=42, stratify=dataset.target)

In [ ]:
from nltk.corpus import stopwords
from nltk import download
download('stopwords')  # Download stopwords list.
stop_words = stopwords.words('english')

def preprocess(sentence):
    return [w for w in sentence.lower().split() if w not in stop_words]

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
doc_embeddings = []
for index,row in df_w2v_train.iterrows():
  #print(row['OriginalTweet'])
  sentence = preprocess(row['text'])
  #print(sentence)
  L = []
  for token in sentence:
    try:
      #print(token,word2vec[token])
      L.append(word2vec[token])
    except:
      1
  if len(L) > 0: tweet_vec = np.mean(np.array(L),axis=0)
  else: tweet_vec = np.zeros(300)
  doc_embeddings.append(tweet_vec)

In [ ]:
X_w2v_train = np.array(doc_embeddings)
y_w2v_train = df_w2v_train.target.to_list()
X_w2v_train.shape

(3901, 300)

In [ ]:
pd.DataFrame(X_w2v_train)

0         1         2    ...       297       298       299
0     0.067671 -0.002075  0.154088  ... -0.013088 -0.020721  0.013532
1     0.009888  0.053017 -0.014893  ... -0.115173  0.066711  0.026489
2    -0.028809  0.020508  0.122314  ... -0.049927 -0.063837  0.000977
3     0.053874  0.004395 -0.021566  ... -0.047526  0.011230  0.012655
4     0.193359 -0.004333 -0.032227  ...  0.052002  0.027588 -0.006592
...        ...       ...       ...  ...       ...       ...       ...
3896 -0.018480  0.035795 -0.023838  ... -0.146679 -0.078510  0.005709
3897  0.089355 -0.010579  0.042155  ... -0.054199  0.070353 -0.022786
3898 -0.015951  0.066447  0.146810  ... -0.195150  0.105632 -0.016276
3899 -0.035812 -0.009743 -0.118286  ... -0.008789 -0.156738 -0.040771
3900  0.125651  0.082316  0.039958  ... -0.143555 -0.020186  0.137288

[3901 rows x 300 columns]

In [ ]:
doc_embeddings = []
for index,row in df_w2v_test.iterrows():
  sentence = preprocess(row['text'])
  L = []
  for token in sentence:
    try:
      #print(token,word2vec[token])
      L.append(word2vec[token])
    except:
      1
  if len(L) > 0: tweet_vec = np.mean(np.array(L),axis=0)
  else: tweet_vec = np.zeros(300)
  doc_embeddings.append(tweet_vec)

In [ ]:
X_w2v_test = np.array(doc_embeddings)
y_w2v_test = df_w2v_test.target.to_list()
X_w2v_test.shape

(1673, 300)

# Extração de padrões (ML)

## Bag of Words

In [ ]:
# Preparando dados para classificador kNN
X = np.array(vsm)
length = np.sqrt((X**2).sum(axis=1))[:,None]
X_bag = X / (length+0.00001)

X.shape

(5574, 7593)

In [ ]:
Y_bag = dataset['target'].to_list()

In [ ]:
X_bag_train, X_bag_test, y_bag_train, y_bag_test = train_test_split(X_bag, Y_bag, test_size=0.30, random_state=42, stratify = Y_bag)

In [ ]:
neigh = KNeighborsClassifier(n_neighbors=3)
neigh.fit(X_bag_train, y_bag_train)

neigh.score(X_bag_test,y_bag_test)

predictions_bag = list(neigh.predict(X_bag_test))

In [ ]:

print('Accuracy score: {}'.format(accuracy_score(y_bag_test, predictions_bag)))
print('Precision score: {}'.format(precision_score(y_bag_test, predictions_bag,pos_label='spam')))
print('Recall score: {}'.format(recall_score(y_bag_test, predictions_bag,pos_label='spam')))
print('F1 score: {}'.format(f1_score(y_bag_test, predictions_bag,pos_label='spam')))

Accuracy score: 0.9282725642558278
Precision score: 1.0
Recall score: 0.4642857142857143
F1 score: 0.6341463414634146


In [ ]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_bag_test,predictions_bag)

array([[1449,    0],
       [ 120,  104]])

## Word2Vec

In [ ]:
neigh = KNeighborsClassifier(n_neighbors=3)
neigh.fit(X_w2v_train, y_w2v_train)

neigh.score(X_w2v_test,y_w2v_test)

predictions_w2v = list(neigh.predict(X_w2v_test))

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
print('Accuracy score: {}'.format(accuracy_score(y_w2v_test, predictions_w2v)))
print('Precision score: {}'.format(precision_score(y_w2v_test, predictions_w2v,pos_label='spam')))
print('Recall score: {}'.format(recall_score(y_w2v_test, predictions_w2v,pos_label='spam')))
print('F1 score: {}'.format(f1_score(y_w2v_test, predictions_w2v,pos_label='spam')))

Accuracy score: 0.9563658099222953
Precision score: 0.8007968127490039
Recall score: 0.8973214285714286
F1 score: 0.8463157894736842


In [ ]:
confusion_matrix(y_w2v_test,predictions_w2v)

array([[1399,   50],
       [  23,  201]])

## BERT

Mesmo obtendo um modelo satisfatório usando Word2Vec, testaremos um modelo que demanda mais poder computacional:

In [ ]:
#!pip install ktrain
import keras
import ktrain
from ktrain import text

In [ ]:
(x_bert_train, y_bert_train), (x_bert_test, y_bert_test), preproc = text.texts_from_df(dataset, 
                                                                   'text',
                                                                   label_columns='target',
                                                                   maxlen=64, 
                                                                   max_features=10000,
                                                                   preprocess_mode='bert',
                                                                   lang=None,
                                                                   val_pct = 0.3,
                                                                   random_state=42
                                                                   )

['ham', 'spam']
      ham  spam
4977  1.0   0.0
650   0.0   1.0
5184  1.0   0.0
708   1.0   0.0
3295  1.0   0.0
['ham', 'spam']
      ham  spam
3690  1.0   0.0
3527  1.0   0.0
724   1.0   0.0
3370  1.0   0.0
468   1.0   0.0
downloading pretrained BERT model (uncased_L-12_H-768_A-12.zip)...
[██████████████████████████████████████████████████]
extracting pretrained BERT model...
done.

cleanup downloaded zip...
done.

preprocessing train...
language: en


Is Multi-Label? False
preprocessing test...
language: en


In [ ]:
model = text.text_classifier('bert', (x_bert_train, y_bert_train) , preproc=preproc)
classifier = ktrain.get_learner(model, 
                             train_data=(x_bert_train, y_bert_train), 
                             val_data=(x_bert_test, y_bert_test),
                             batch_size=64
                             )

Is Multi-Label? False
maxlen is 64
done.


In [ ]:
classifier.fit_onecycle(0.00002,2)



begin training using onecycle policy with max lr of 2e-05...
Epoch 1/5
61/61 [==============================] - 3008s 49s/step - loss: 0.4402 - accuracy: 0.7765 - val_loss: 0.1162 - val_accuracy: 0.9809
Epoch 2/5
61/61 [==============================] - 2934s 48s/step - loss: 0.0542 - accuracy: 0.9859 - val_loss: 0.0286 - val_accuracy: 0.9934
Epoch 3/5
22/61 [=========>....................] - ETA: 27:59 - loss: 0.0248 - accuracy: 0.9922

KeyboardInterrupt: ignored

In [ ]:
resul = classifier.validate()
resul

              precision    recall  f1-score   support

           0       1.00      1.00      1.00      1447
           1       0.99      0.97      0.98       226

    accuracy                           0.99      1673
   macro avg       0.99      0.98      0.99      1673
weighted avg       0.99      0.99      0.99      1673



array([[1445,    2],
       [   7,  219]])

In [ ]:
classifier_model = ktrain.get_predictor(classifier.model, preproc)

In [ ]:
classifier_model.save('/content/drive/MyDrive/graduação/mineracao_dados_nao_estruturados/my_model')

/usr/local/lib/python3.7/dist-packages/keras/utils/generic_utils.py:497: CustomMaskWarning:

Custom mask layers require a config and must override get_config. When loading, the custom mask layer must be passed to the custom_objects argument.



# Pós - processamento

- Acurácia: indica uma performance geral do modelo. Dentre todas as classificações, quantas o modelo classificou corretamente;

- Precisão: dentre todas as classificações de classe Positivo que o modelo fez, quantas estão corretas;

- Recall: dentre todas as situações de classe Positivo como valor esperado, quantas estão corretas;

- F1-Score: média harmônica entre precisão e recall.

Ao receber spams em SMS, é comum que eles contenham vírus ou busquem informações confidenciais do usuário. Sendo assim, o melhor modelo seria aquele que classifique corretamente spams o máximo possível, mesmo que para isso ele sugira que mensagens normais sejam spams também.

Neste caso, a melhor métrica para avaliar seria Recall

In [ ]:
d_2 = { " ": ['Acurácia', 'Precisão', 'Recall', 'F1 score'],
    "Modelo Bag of Words": [accuracy_score(y_bag_test, predictions_bag),
                            precision_score(y_bag_test, predictions_bag,pos_label='spam'),
                            recall_score(y_bag_test, predictions_bag,pos_label='spam'),
                            f1_score(y_bag_test, predictions_bag,pos_label='spam')],
    "Modelo Word2Vec": [accuracy_score(y_w2v_test, predictions_w2v),
                        precision_score(y_w2v_test, predictions_w2v,pos_label='spam'),
                        recall_score(y_w2v_test, predictions_w2v,pos_label='spam'),
                        f1_score(y_w2v_test, predictions_w2v,pos_label='spam')]
}
df_2 = pd.DataFrame(data = d_2)
df_2

Modelo Bag of Words  Modelo Word2Vec
0  Acurácia             0.928273         0.956366
1  Precisão             1.000000         0.800797
2    Recall             0.464286         0.897321
3  F1 score             0.634146         0.846316

In [ ]:
confusion_matrix(y_bag_test,predictions_bag)

array([[1449,    0],
       [ 120,  104]])

In [ ]:
confusion_matrix(y_w2v_test,predictions_w2v)

array([[1399,   50],
       [  23,  201]])

O modelo usando Bag of Words, apesar de ter uma boa acurácia, não foi um bom modelo para o nosso problema, devido ao seu baixo Recall.

Fazendo uma comparação, o modelo Bag of Words teve 120 casos de falso negativo, enquanto o Word2Vec teve 23 casos, em um total de 1673 casos de teste. 

Considerando uma cidade com uma população de 100 mil habitantes que recebem um SMS por dia, teríamos 7173 pessoas recebendo Spam sem serem alertados no primeiro modelo, já no segundo teríamos 1375. 

Por fim, o vencedor seria o modelo Word2Vector.

Este resultado consideou também o custo computacional. A base utilizada possui menos de 6000 SMS, porém, no "mundo real" teríamos um a base com uma quantidade muito superior de dados.

Deixando de lado o custo computacional, utilizando o modelo BERT, tivemos resultados muito superiores

In [ ]:
resul

              precision    recall  f1-score   support

           0       1.00      1.00      1.00      1447
           1       0.99      0.97      0.98       226

    accuracy                           0.99      1673
   macro avg       0.99      0.98      0.99      1673
weighted avg       0.99      0.99      0.99      1673



array([[1445,    2],
       [   7,  219]])

Usando o exemplo anterior da cidade com 100 mil habitantes, com o BERT teríamos 418 recebendo Spam sem alerta, um valor 3 vezes menor que o Word2Vec e 17 vezes menor que o Bag of Words

# Uso do conhecimento

Neste trabalho extraimos padrões de um conjunto de SMS, sendo assim, a aplicação mais óbvia é implementar o modelo no aplicativo de mensagens do usuário para que ele seja notificado caso aquele SMS tenha características de Spam, porém, isso também poderia ser feito com emails e caixas de mensagens em redes sociais como facebook e instagram.

Refazer: : Em relação ao pre-processamneto, poderia ter variado o max_features da BoW e utilizar o tfidf. Em relação a classificação, poderia ter variado alguns parâmetros do método (número de vizinhos no kNN) e utilizar outros métodos interessantes para classificação para compara-los. No mundo real, pode ser mais facil para o usuario identificar Spams no meio de mensagem
reais que ele vai olhar com calma porque são importantes, em vez de procurar pro mensagens reais no meio de varias Spams (mensagens
nao importantes)? Considerar a F1-Score invés do recall? 